File: tis_agent/time_saved.py
"""Time saved KPI helpers — keep in sync with admin/lib/time-saved.ts."""

from __future__ import annotations

DEFAULT_MINUTES_SAVED_PER_QUESTION = 5
MIN_MINUTES_SAVED_PER_QUESTION = 1
MAX_MINUTES_SAVED_PER_QUESTION = 180


def is_tina_handled(outcome: str, human_replied_at: str | None = None) -> bool:
    """True when interactions.outcome is a grounded Tina success with no human reply."""
    return outcome == "success" and not human_replied_at


def clamp_minutes_per_question(raw: object) -> int:
    try:
        n = round(float(raw))  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return DEFAULT_MINUTES_SAVED_PER_QUESTION
    return min(MAX_MINUTES_SAVED_PER_QUESTION, max(MIN_MINUTES_SAVED_PER_QUESTION, n))


def format_saved_time(total_minutes: float) -> str:
    minutes = max(0, round(total_minutes))
    if minutes < 60:
        return f"{minutes}m"
    hours, rest = divmod(minutes, 60)
    if rest == 0:
        return f"{hours}h"
    return f"{hours}h {rest}m"


def time_saved_minutes(handled_count: int, minutes_per_question: object) -> int:
    return max(0, int(handled_count)) * clamp_minutes_per_question(minutes_per_question)


def format_saved_time_delta(current_minutes: float, previous_minutes: float) -> str | None:
    current = max(0, round(current_minutes))
    previous = max(0, round(previous_minutes))
    if previous == 0:
        if current == 0:
            return "→ 0m"
        return f"↑ {format_saved_time(current)}"
    diff = current - previous
    if diff == 0:
        return "→ 0m"
    arrow = "↑" if diff > 0 else "↓"
    return f"{arrow} {format_saved_time(abs(diff))}"

File: admin/lib/time-saved.ts
/** Time saved KPI: Tina successes × configured minutes per question. */

export const DEFAULT_MINUTES_SAVED_PER_QUESTION = 5;
export const MIN_MINUTES_SAVED_PER_QUESTION = 1;
export const MAX_MINUTES_SAVED_PER_QUESTION = 180;

export const TIME_SAVED_DEFINITION =
  "Estimated time saved based on questions successfully handled by Tina multiplied by the configured manual handling time per question.";

/** Grounded Tina answers (`interactions.outcome = success`) with no human reply. */
export function isTinaHandled(
  outcome: string,
  humanRepliedAt?: string | null,
): boolean {
  return outcome === "success" && !humanRepliedAt;
}

export function clampMinutesPerQuestion(raw: unknown): number {
  const n = typeof raw === "number" ? raw : Number(raw);
  if (!Number.isFinite(n)) return DEFAULT_MINUTES_SAVED_PER_QUESTION;
  return Math.min(
    MAX_MINUTES_SAVED_PER_QUESTION,
    Math.max(MIN_MINUTES_SAVED_PER_QUESTION, Math.round(n)),
  );
}

export function formatSavedTime(totalMinutes: number): string {
  const minutes = Math.max(0, Math.round(totalMinutes));
  if (minutes < 60) return `${minutes}m`;
  const hours = Math.floor(minutes / 60);
  const rest = minutes % 60;
  if (rest === 0) return `${hours}h`;
  return `${hours}h ${rest}m`;
}

export function timeSavedMinutes(
  handledCount: number,
  minutesPerQuestion: number,
): number {
  return Math.max(0, Math.round(handledCount)) * clampMinutesPerQuestion(minutesPerQuestion);
}

export function formatSavedTimeDelta(
  currentMinutes: number,
  previousMinutes: number,
): string | null {
  const current = Math.max(0, Math.round(currentMinutes));
  const previous = Math.max(0, Math.round(previousMinutes));
  if (previous === 0) {
    if (current === 0) return "→ 0m";
    return `↑ ${formatSavedTime(current)}`;
  }
  const diff = current - previous;
  if (diff === 0) return "→ 0m";
  const arrow = diff > 0 ? "↑" : "↓";
  return `${arrow} ${formatSavedTime(Math.abs(diff))}`;
}
